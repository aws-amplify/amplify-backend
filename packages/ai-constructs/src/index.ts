import * as conversation from './conversation';

export { conversation };
