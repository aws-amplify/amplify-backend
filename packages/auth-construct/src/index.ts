export {
  AuthProps,
  EmailLogin,
  EmailLoginSettings,
  GoogleProviderProps,
  FacebookProviderProps,
  AmazonProviderProps,
  AppleProviderProps,
  OidcProviderProps,
  SamlProviderProps,
  ExternalProviderOptions,
  VerificationEmailWithCode,
  VerificationEmailWithLink,
  MFA,
  MFASettings,
  PhoneNumberLogin,
  TriggerEvent,
  IdentityProviderProps,
  AttributeMapping,
} from './types.js';
export { AmplifyAuth } from './construct.js';
export { triggerEvents } from './trigger_events.js';
