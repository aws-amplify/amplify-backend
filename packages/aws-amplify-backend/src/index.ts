export * from '@aws-amplify/backend';
