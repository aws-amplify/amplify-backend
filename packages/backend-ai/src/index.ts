export * from './types.js';

export { getMessage } from './get_message.js';
