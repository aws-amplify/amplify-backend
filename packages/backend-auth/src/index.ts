export { BackendAuth, AmplifyAuthProps, defineAuth } from './factory.js';
export * from './types.js';
