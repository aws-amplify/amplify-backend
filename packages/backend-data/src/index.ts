export { defineData } from './factory.js';
export {
  ApiKeyAuthorizationModeProps,
  LambdaAuthorizationModeProps,
  OIDCAuthorizationModeProps,
  DefaultAuthorizationMode,
  AuthorizationModes,
  DataSchemaInput,
  DataProps,
} from './types.js';
