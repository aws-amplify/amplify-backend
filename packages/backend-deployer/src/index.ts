export * from './cdk_deployer_singleton_factory.js';
export * from './types.js';
