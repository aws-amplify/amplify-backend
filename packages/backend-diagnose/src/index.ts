export { getCdkInfo, formatCdkInfo } from './cdk_info.js';
export { getEnvInfo, formatEnvInfo } from './env_info.js';
