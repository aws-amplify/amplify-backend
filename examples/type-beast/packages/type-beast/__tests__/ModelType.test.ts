import { expectTypeTestsToPassAsync } from 'jest-tsd';
import a, { defineData, ClientSchema } from '../index';
import { schemaPreprocessor } from '../src/SchemaProcessor';
import {
  PublicProviders,
  PrivateProviders,
  Operations,
  Operation,
} from '../src/Authorization';

// evaluates type defs in corresponding test-d.ts file
it('should not produce static type errors', async () => {
  await expectTypeTestsToPassAsync(__filename);
});

describe('model auth rules', () => {
  // it('rejects items that are not auth rules', () => {

  // });

  it('can define public auth with no provider', () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([a.allow.public()]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it('fails loudly on invalid provider', () => {
    expect(() => {
      a.schema({
        widget: a
          .model({
            title: a.string(),
          })
          // @ts-expect-error
          .authorization([a.allow.public('bad-provider')]),
      });
    }).toThrow();
  });

  it('can define private auth with no provider', () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([a.allow.private()]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it('can define owner auth with no provider', () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([a.allow.owner()]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it('can define owner auth with owner field spec', () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([a.allow.owner().inField('title')]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`can specify operations `, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([a.allow.public().to(['create', 'read'])]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`can specify an owner identityClaim`, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([a.allow.owner().identityClaim('user_id')]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`can create a "multiple owners" rule on an existing field`, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
          authors: a.string().array(),
        })
        .authorization([a.allow.multipleOwners().inField('authors')]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`can create a "multiple owners" rule an on implied (auto-created) field`, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([a.allow.multipleOwners().inField('authors')]),
    });

    // what customers export
    type Schema = ClientSchema<typeof schema>;

    // If the types are correct, this field should exist
    type AuthorsType = Schema['widget']['authors'];

    // I should NOT be able to assign a singular `string` to it
    // @ts-expect-error
    let authors: AuthorsType = 'string';

    // I SHOULD be able to assign a `string[]`;
    authors = ['username1', 'username2'];

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`throws a fit of an implied (auto-created) field conflicts with an explicit field`, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),

          // pluralized `author` fields
          author: a.string().array(),
        })

        // authorization expects `author` to be singular `string`
        // @ts-expect-error
        .authorization([a.allow.owner().inField('author')]),
    });
  });

  it(`can create a static Admins group rule`, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([a.allow.specificGroup('Admins')]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`can create a static [Admins, Moderators] groups rule`, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([a.allow.specificGroups(['Admins', 'Moderators'])]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`can create a dynamic singular groups rule`, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([a.allow.groupDefinedIn('businessUnitOwner')]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`can create a dynamic multi groups rule`, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([
          a.allow.groupsDefinedIn('sharedWithGroups').to(['read']),
        ]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`can chain off of inField rules`, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([
          a.allow.owner().inField('customOwnerField').to(['create', 'read']),
        ]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`can chain off of multi-owner inField rules`, () => {
    const schema = a.schema({
      widget: a
        .model({
          title: a.string(),
        })
        .authorization([
          a.allow
            .multipleOwners()
            .inField('customOwnerField')
            .to(['create', 'read']),
        ]),
    });

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`includes auth from fields`, () => {
    const schema = a.schema({
      widget: a.model({
        id: a.id(),
        title: a
          .string()
          .authorization([
            a.allow.owner().inField('customOwner').to(['create', 'read']),
          ]),
      }),
    });

    type Schema = ClientSchema<typeof schema>;
    type CustomOwnerType = Schema['widget']['customOwner'];

    // single owner should be allowed
    let customOwner: CustomOwnerType = 'abc';

    // multiple owners should not
    // @ts-expect-error
    customOwner = ['multiple', 'owners'];

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  it(`includes auth from related model fields`, () => {
    const schema = a.schema({
      widget: a.model({
        id: a.id(),
        parent: a
          .belongsTo('widget')
          .authorization([
            a.allow.owner().inField('customOwner').to(['create', 'read']),
          ]),
      }),
    });

    type Schema = ClientSchema<typeof schema>;
    type CustomOwnerType = Schema['widget']['customOwner'];

    // single owner should be allowed
    let customOwner: CustomOwnerType = 'abc';

    // multiple owners should not
    // @ts-expect-error
    customOwner = ['multiple', 'owners'];

    const graphql = schemaPreprocessor(schema).processedSchema;
    expect(graphql).toMatchSnapshot();
  });

  for (const provider of PublicProviders) {
    it(`can define public with with provider ${provider}`, () => {
      const schema = a.schema({
        widget: a
          .model({
            title: a.string(),
          })
          .authorization([a.allow.public(provider)]),
      });

      const graphql = schemaPreprocessor(schema).processedSchema;
      expect(graphql).toMatchSnapshot();
    });

    const TestOperations: Operation[][] = [
      // each individual operation
      ...Operations.map((op) => [op]),

      // a couple sanity checks to support a combinations
      ['create', 'read', 'update', 'delete'],
      ['create', 'read', 'listen'],
    ];

    for (const operations of TestOperations) {
      it(`can define public with with provider ${provider} for operations ${operations}`, () => {
        const schema = a.schema({
          widget: a
            .model({
              title: a.string(),
            })
            .authorization([
              a.allow.public(provider).to(operations),
              // 'whatever',
            ]),
        });

        const graphql = schemaPreprocessor(schema).processedSchema;
        expect(graphql).toMatchSnapshot();
      });
    }
  }

  for (const provider of PrivateProviders) {
    it(`can define private with with provider ${provider}`, () => {
      const schema = a.schema({
        widget: a
          .model({
            title: a.string(),
          })
          .authorization([a.allow.private(provider)]),
      });

      const graphql = schemaPreprocessor(schema).processedSchema;
      expect(graphql).toMatchSnapshot();
    });

    const TestOperations: Operation[][] = [
      // each individual operation
      ...Operations.map((op) => [op]),

      // a couple sanity checks to support a combinations
      ['create', 'read', 'update', 'delete'],
      ['create', 'read', 'listen'],
    ];

    for (const operations of TestOperations) {
      it(`can define private with with provider ${provider} for operations ${operations}`, () => {
        const schema = a.schema({
          widget: a
            .model({
              title: a.string(),
            })
            .authorization([a.allow.private(provider).to(operations)]),
        });

        const graphql = schemaPreprocessor(schema).processedSchema;
        expect(graphql).toMatchSnapshot();
      });
    }
  }
});
