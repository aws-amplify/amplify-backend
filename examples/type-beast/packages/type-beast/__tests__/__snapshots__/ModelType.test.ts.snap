// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model auth rules can chain off of inField rules 1`] = `
"type widget @model @auth(rules: [{allow: owner, operations: [create, read], ownerField: "customOwnerField"}])
{
  customOwnerField: String
  title: String!
}"
`;

exports[`model auth rules can chain off of multi-owner inField rules 1`] = `
"type widget @model @auth(rules: [{allow: owner, operations: [create, read], ownerField: "customOwnerField"}])
{
  customOwnerField: [String!]
  title: String!
}"
`;

exports[`model auth rules can create a "multiple owners" rule an on implied (auto-created) field 1`] = `
"type widget @model @auth(rules: [{allow: owner, ownerField: "authors"}])
{
  authors: [String!]
  title: String!
}"
`;

exports[`model auth rules can create a "multiple owners" rule on an existing field 1`] = `
"type widget @model @auth(rules: [{allow: owner, ownerField: "authors"}])
{
  authors: [String!]!
  title: String!
}"
`;

exports[`model auth rules can create a dynamic multi groups rule 1`] = `
"type widget @model @auth(rules: [{allow: groups, operations: [read], groupsField: "sharedWithGroups"}])
{
  sharedWithGroups: [String!]
  title: String!
}"
`;

exports[`model auth rules can create a dynamic singular groups rule 1`] = `
"type widget @model @auth(rules: [{allow: groups, groupsField: "businessUnitOwner"}])
{
  businessUnitOwner: String
  title: String!
}"
`;

exports[`model auth rules can create a static [Admins, Moderators] groups rule 1`] = `
"type widget @model @auth(rules: [{allow: groups, groups: ["Admins", "Moderators"]}])
{
  title: String!
}"
`;

exports[`model auth rules can create a static Admins group rule 1`] = `
"type widget @model @auth(rules: [{allow: groups, groups: ["Admins"]}])
{
  title: String!
}"
`;

exports[`model auth rules can define owner auth with no provider 1`] = `
"type widget @model @auth(rules: [{allow: owner}])
{
  title: String!
}"
`;

exports[`model auth rules can define owner auth with owner field spec 1`] = `
"type widget @model @auth(rules: [{allow: owner, ownerField: "title"}])
{
  title: String!
}"
`;

exports[`model auth rules can define private auth with no provider 1`] = `
"type widget @model @auth(rules: [{allow: private}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations create 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [create]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations create,read,listen 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [create, read, listen]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations create,read,update,delete 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [create, read, update, delete]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations delete 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [delete]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations get 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [get]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations list 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [list]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations listen 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [listen]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations read 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [read]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations search 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [search]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations sync 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [sync]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider iam for operations update 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: iam, operations: [update]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations create 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [create]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations create,read,listen 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [create, read, listen]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations create,read,update,delete 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [create, read, update, delete]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations delete 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [delete]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations get 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [get]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations list 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [list]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations listen 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [listen]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations read 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [read]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations search 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [search]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations sync 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [sync]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider oidc for operations update 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: oidc, operations: [update]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations create 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [create]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations create,read,listen 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [create, read, listen]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations create,read,update,delete 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [create, read, update, delete]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations delete 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [delete]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations get 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [get]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations list 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [list]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations listen 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [listen]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations read 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [read]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations search 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [search]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations sync 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [sync]}])
{
  title: String!
}"
`;

exports[`model auth rules can define private with with provider userPools for operations update 1`] = `
"type widget @model @auth(rules: [{allow: private, provider: userPools, operations: [update]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public auth with no provider 1`] = `
"type widget @model @auth(rules: [{allow: public}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations create 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [create]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations create,read,listen 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [create, read, listen]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations create,read,update,delete 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [create, read, update, delete]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations delete 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [delete]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations get 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [get]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations list 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [list]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations listen 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [listen]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations read 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [read]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations search 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [search]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations sync 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [sync]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider apiKey for operations update 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: apiKey, operations: [update]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations create 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [create]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations create,read,listen 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [create, read, listen]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations create,read,update,delete 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [create, read, update, delete]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations delete 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [delete]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations get 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [get]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations list 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [list]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations listen 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [listen]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations read 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [read]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations search 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [search]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations sync 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [sync]}])
{
  title: String!
}"
`;

exports[`model auth rules can define public with with provider iam for operations update 1`] = `
"type widget @model @auth(rules: [{allow: public, provider: iam, operations: [update]}])
{
  title: String!
}"
`;

exports[`model auth rules can specify an owner identityClaim 1`] = `
"type widget @model @auth(rules: [{allow: owner, identityClaim: "user_id"}])
{
  title: String!
}"
`;

exports[`model auth rules can specify operations  1`] = `
"type widget @model @auth(rules: [{allow: public, operations: [create, read]}])
{
  title: String!
}"
`;
